### PR TITLE
Establish compatibility of typescript block with webpack v4 - fixes #296

### LINKED ISSUE #296: Update typescript block to work with webpack 4
I'm using webpack-blocks to provide configuration to react-styleguidist and experienced the following errors.

```bash
TypeError: resolver.ensureHook is not a function
TypeError: resolver.ensureHook is not a function
    at ModulesInRootPlugin.apply (C:\Users\sh4dow\development\github\fructosedb\node_modules\enhanced-resolve\lib\ModulesInRootPlugin.js:15:27)
    at PathPlugin.apply (C:\Users\sh4dow\development\github\fructosedb\node_modules\awesome-typescript-loader\src\paths-plugin.ts:114:71)
    at webpack (C:\Users\sh4dow\development\github\fructosedb\node_modules\webpack\lib\webpack.js:37:12)
    at createServer (C:\Users\sh4dow\development\github\fructosedb\node_modules\react-styleguidist\scripts\create-server.js:27:19)
    at server (C:\Users\sh4dow\development\github\fructosedb\node_modules\react-styleguidist\scripts\server.js:5:21)
    at commandServer (C:\Users\sh4dow\development\github\fructosedb\node_modules\react-styleguidist\bin\styleguidist.js:123:19)
    at Object.<anonymous> (C:\Users\sh4dow\development\github\fructosedb\node_modules\react-styleguidist\bin\styleguidist.js:72:3)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
    at startup (internal/bootstrap/node.js:266:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:596:3)
error Command failed with exit code 1.
```

I've resolved it locally by using [this configuration](https://github.com/s-panferov/awesome-typescript-loader/issues/534#issuecomment-394368316) and updating the versions of awesome-typescript-loader and typescript to the latest ones.  (The latter just to get rid of most deprecation warnings)

If you want me to, I'll prepare a PR for this.

### PR DESCRIPTION
**Pull request by s-h-a-d-o-w**
While preparing this, I of course took a closer look at what the configuration I've based this on (see: https://github.com/s-panferov/awesome-typescript-loader/issues/534#issuecomment-394368316 ) actually does and whether/why they are necessary - details see below.
Obviously, just let me know if you want something changed. Like - upgrading TS to v3 may not be necessary but with this being part of a new major release anyway, I figured it would make sense.
Oh yeah and I'm not familiar with your tests and what the snapshots should look like, so why merely updating the snapshots added thousands of LOC and whether that's good/bad/neutral... don't know.

1. Moving `TsConfigPathsPlugin` to resolve gets rid of `resolver.ensureHook is not a function`
But leads to `Unexpected token` possibly because of JSX
2. `Unexpected token` is resolved by `useBabel: true`
3. Since I guess the point of webpack-blocks is reasonable defaults, I've included these nonessential options:
- `useCache: true` with `cacheDirectory` set to node_modules, like `babel-loader` does (Why: https://github.com/s-panferov/awesome-typescript-loader#differences-between-ts-loader)
- `babelCore: ...` I have babel bridge installed but maybe not everyone does (https://github.com/s-panferov/awesome-typescript-loader#babelcore-string-defaultundefined)
- `babelOptions: ...` Sensible quality of life and bundle size reduction

### CODE REVIEW

**Comment by s-h-a-d-o-w** [comment]
Just saw that CI failed - help with that would be appreciated because I have no clue why `PathPlugin` expands to this huge object when `TsConfigPathsPlugin` is moved to resolve.

File: packages/typescript/index.js [hunk 1]
```diff
@@ -26,17 +30,26 @@ function typescript(options = {}) {
               use: [
                 {
                   loader: 'awesome-typescript-loader',
-                  options
+                  options: Object.assign(
+                    {
+                      useCache: true,
+                      cacheDirectory: 'node_modules/.awcache',
+                      useBabel: true,
```
Comment (vlad-zhukov) [inline packages/typescript/index.js]: Let's remove `babel`-related settings from the defaults.

Comment (vlad-zhukov) [reply packages/typescript/index.js]: `useBabel` should be remove as well

Comment (s-h-a-d-o-w) [reply packages/typescript/index.js]: Is it a given that typescript block users are aware that they will need to pass these settings themselves if they use JSX? Or is it documented somewhere already?

Comment (s-h-a-d-o-w) [reply packages/typescript/index.js]: Never mind, I just saw that the sample app contains a .tsx file anyway and the test passes. So I guess there's simply something wrong with the webpack blocks config in my own app.

**Comment by s-h-a-d-o-w** [comment]
Hm... alright, I don't know what's going on here because you're already testing tsx with that sample-app and it passes. I didn't see that before I started working on this.

Why I get a TypeError in my own app and others using awesome-typescript-loader as well... who knows. But I guess this PR is simply not necessary.

File: packages/typescript/index.js (hunk 1 above)
Comment (vlad-zhukov) [reply packages/typescript/index.js]: TypeScript supports JSX out of the box.

**Comment by jvanbruegge** [comment]
You need to configure JSX in your tsconfig, do you have that?

**Comment by s-h-a-d-o-w** [comment]
Well... it was set "preserve". Setting it to "react" did the trick for problem 2 and `useBabel` is no longer required.

Thanks for making me aware of this!
But the major question (that TypeError) unfortunately remains.

**Comment by andywer** [comment]
Thanks for submitting the PR! :)

Yeah, something is severely wrong about that test snapshot... 😅 I will have a look at that.

Thanks everyone for reviewing so quickly. I am terribly busy ATM, but I will find some time to help getting this baby merged 🙌

**Comment by s-h-a-d-o-w** [comment]
I already addressed that, just didn't push it because I assumed that my changes are useless after all.
But... I just read the docs for the paths plugin and found this **very** interesting bit:

[> Notice that the plugin is placed in the resolve.plugins section of the configuration.](https://github.com/dividab/tsconfig-paths-webpack-plugin#how-to-use)

So... it seems that this PR actually does address a valid issue. (It of course still begs the question why there haven't been issues with it in the past...)

**Comment by jvanbruegge** [comment]
I assume because noone used the path resolving. When I created the typescript block, I had no idea what it was or how to set it up, so the result was a best efford

**Comment by s-h-a-d-o-w** [comment]
Ah I see. Same here. ;)

While trying to learn more about this, I found that the Path Plugin actually doesn't take arguments for `tsconfig` and `compiler`: https://github.com/dividab/tsconfig-paths-webpack-plugin#options
Just removed them and tests still pass locally.

Also: It seems that `{ configFileName: path.resolve(__dirname, './tsconfig.json') }` could be removed from webpack.config.babel.js (in sample-app), since it uses the same path either way.

Let me know if you would like me to include these changes in this PR.

**Comment by s-h-a-d-o-w** [comment]
Brain fart, please disregard my now deleted last comment. ;)

**Comment by andywer** [comment]
Any reason not to merge this now? 😉

**Comment by andywer** [comment]
What's the status here? 🙈

**Comment by jvanbruegge** [comment]
AFAICT I am using the newest webpack-blocks version with webpack 4 without issues, can check the exact version at home though

**Comment by andywer** [comment]
Stale and old. Dropped the ball on that one 🙃

Closing for the time being.

**Pull request closed without merging**